parsers/bibtex.c

```diff
@@ -0,0 +1,459 @@
+		tagEntryInfo e;
+		initTagEntry (&e, name, kind);
+
+		e.lineNumber   = token->lineNumber;
+		e.filePosition = token->filePosition;
+
+		makeTagEntry (&e);
+		vStringDelete (parentName);
+	}
+}
+
+/*
+ *	 Parsing functions
+ */
+
+/*
+ *	Read a C identifier beginning with "firstChar" and places it into
+ *	"name".
+ */
+static void parseIdentifier (vString *const string, const int firstChar)
+{
+	int c = firstChar;
+	Assert (isIdentChar (c));
+	do
+	{
+		vStringPut (string, c);
+		c = getcFromInputFile ();
+	} while (c != EOF && isIdentChar (c));
+	if (c != EOF)
+		ungetcToInputFile (c);		/* unget non-identifier character */
+}
+
+static bool readTokenFull (tokenInfo *const token, const bool includeWhitespaces)
+{
+	int c;
+	int whitespaces = -1;
+
+	token->type			= TOKEN_UNDEFINED;
+	token->keyword		= KEYWORD_NONE;
+	vStringClear (token->string);
+
+getNextChar:
+
+	do
+	{
+		c = getcFromInputFile ();
+		whitespaces++;
+	}
+	while (c == '\t'  ||  c == ' ' ||  c == '\n');
+
+	token->lineNumber   = getInputLineNumber ();
+	token->filePosition = getInputFilePosition ();
+
+	if (includeWhitespaces && whitespaces > 0 && c != '%' && c != EOF)
+	{
+		ungetcToInputFile (c);
+		c = ' ';
+	}
+
+	token->type = (unsigned char) c;
+	switch (c)
+	{
+		case EOF: return false;
+
+		case '@':
+				  /*
+				   * All Bib entries start with an at symbol.
+				   * Check if the next character is an alpha character
+				   * else it is not a potential tex tag.
+				   */
+				  c = getcFromInputFile ();
+				  if (! isalpha (c))
+					  ungetcToInputFile (c);
+				  else
+				  {
+					  vStringPut (token->string, '@');
+					  parseIdentifier (token->string, c);
+					  token->keyword = lookupKeyword (vStringValue (token->string) + 1, Lang_bib);
+					  if (isKeyword (token, KEYWORD_NONE)){
+						  token->type = TOKEN_IDENTIFIER;
+					  }else{
+						  token->type = TOKEN_KEYWORD;
+            }
+				  }
+				  break;
+		case '%':
+				  skipToCharacterInInputFile ('\n'); /* % are single line comments */
+				  goto getNextChar;
+				  break;
+
+		default:
+				  if (isIdentChar (c))
+				  {
+					  parseIdentifier (token->string, c);
+            token->type = TOKEN_IDENTIFIER;
+				  }
+				  break;
+	}
+	return true;
+}
+
+static bool readToken (tokenInfo *const token)
+{
+	return readTokenFull (token, false);
+}
+
+static void copyToken (tokenInfo *const dest, tokenInfo *const src)
+{
+	dest->lineNumber = src->lineNumber;
+	dest->filePosition = src->filePosition;
+	dest->type = src->type;
+	dest->keyword = src->keyword;
+	vStringCopy (dest->string, src->string);
+	vStringCopy (dest->scope, src->scope);
+}
+
+/*
+ *	 Scanning functions
+ */
+
+static bool parseTag (tokenInfo *const token, bibKind kind)
+{
+	tokenInfo *const name = newToken ();
+	vString *	currentid;
+	bool		useLongName = true;
+	bool        eof = false;
+
+	currentid = vStringNew ();
+	/*
+	 * Bib entries are of these formats:
+	 *   @article{identifier,
+   *   author="John Doe"}
```

Comment:
It seems that indentation is a bit broken.

Reply:
tabs and whitespaces got confused in my editor... I tried to fix indentation in 952f651